File: python/map_service.py
import datetime
from dataclasses import dataclass
from math import sin, cos, sqrt, atan2, radians

# approximate radius of earth in km
R = 6373.0
MINUTES_PER_HOUR = 60
SECONDS_PER_HOUR = 3600

@dataclass
class Location:
    latitude: float
    longitude: float

class MapService:
    def __init__(self, average_speed=50):
        "average speed in km/h"
        self.average_speed = average_speed

    def calculate_eta(self, location1 : Location, location2 : Location) -> float:
        "return the number of minutes it will take to travel between locations at average speed."
        distance = self.calculate_distance(location1, location2)
        return distance / self.average_speed * MINUTES_PER_HOUR

    def calculate_distance(self, location1 : Location, location2 : Location) -> float:
        lat1 = radians(location1.latitude)
        lon1 = radians(location1.longitude)
        lat2 = radians(location2.latitude)
        lon2 = radians(location2.longitude)
        dlon = lon2 - lon1
        dlat = lat2 - lat1
        a = sin(dlat / 2) ** 2 + cos(lat1) * cos(lat2) * sin(dlon / 2) ** 2
        c = 2 * atan2(sqrt(a), sqrt(1 - a))
        # km
        distance = R * c
        return distance

    def update_average_speed(self, location1 : Location, location2 : Location, elapsed_time: datetime.timedelta):
        distance = self.calculate_distance(location1, location2)
        updated_speed = distance / (elapsed_time.seconds / SECONDS_PER_HOUR)
        self.average_speed = updated_speed

File: typescript/src/mapService.ts

const MINUTES_PER_HOUR = 60;
const SECONDS_PER_HOUR = 60 * MINUTES_PER_HOUR;
const R = 6373.0; // Radius of the earth

const radians = (d: number) : number => d * (Math.PI / 180);

export interface Location {
    latitude: number
    longitude: number
}

export class MapService {

    #averageSpeed : number;

    public calculateETA(location1: Location, location2: Location) {
        const distance = this.calculateDistance(location1, location2)
        return distance / this.#averageSpeed * MINUTES_PER_HOUR;
    }

    public calculateDistance(location1: Location, location2: Location) {
        const lat1 = radians(location1.latitude);
        const lon1 = radians(location1.longitude);
        const lat2 = radians(location2.latitude);
        const lon2 = radians(location2.longitude);

        const dlat = lat2 - lat1
        const dlon = lon2 - lon1

        const a = Math.sin(dlat / 2) ** 2 + Math.cos(lat1) * Math.cos(lat2) * Math.sin(dlon / 2) ** 2
        const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));

        return R * c;
    }

    public updateAverageSpeed(location1: Location, location2: Location, elapsed: number) {
        const distance = this.calculateDistance(location1, location2)
        this.#averageSpeed = distance / (elapsed / SECONDS_PER_HOUR)
    }

}
